remove make-gate-matrix-from-gate-string in favour of make-matrix-from-quil

notmgsk opened the pull request:
closes #172

ecpeterson commented on src/compressor/compressor.lisp:
```diff
@@ -404,7 +404,9 @@ other's."
   "This routine is called by COMPRESS-INSTRUCTIONS-IN-CONTEXT to make a decision about how to prefer 'linear algebraic compression': the list of INSTRUCTIONS can always be rewritten as its associated action matrix, but under certain conditions (governed by CONTEXT) we can sometimes get away with something less."
-  (let ((qubits-on-obj (qubits-in-instr-list instructions)))
+  (let* ((qubits-on-obj (qubits-in-instr-list instructions))
+         (qubits-indices (mapcar (alexandria:compose #'qubit-index #'qubit)
```
aren't these functions mutual inverses?

stylewarning replied:
yes; if the intention here is to copy the list, we should call `copy-list`

notmgsk replied:
yep. thank McCarthy for code review 🙏

ecpeterson commented on src/compilers/optimal-2q.lisp:
```diff
@@ -385,8 +385,9 @@ The optional argument INSTR is used to canonicalize the qubit indices of the ins
-               (let ((temp (make-gate-matrix-from-gate-string (application-arguments instr)
-                                                              bare-circuit)))
+               (let ((temp (make-matrix-from-quil bare-circuit
+                                                  :relabeling (mapcar #'qubit-index
```
I think you also mean to use your helper function `qubits-relabeling` here; otherwise, this doesn't seem to be an alist.

notmgsk replied:
yep. thank McCarthy for code review 🙏

stylewarning replied:
@notmgsk this line needs changing still

stylewarning commented on src/gates.lisp:
```diff
@@ -569,8 +569,16 @@ Note that this is a controlled version of a R_z gate multiplied by a phase."
                   m)))
       (magicl:multiply-complex-matrices mat m))))
 
+(defun qubits-relabeling (qubits-list)
```
How about we use functions instead to represent these maps? I suggest rewriting this as something like this:
```
(defun standard-qubit-relabeler (qubits-list)
  "Construct a standard qubit relabeler and its inverse.

A *relabeler* is a function that takes as input a logical qubit index and produces a relabeled (sometimes called *physical*) qubit index.

The *standard labeling* of qubits (in a context where qubits are supplied as arguments to a gate application) is in descending order to 0. As such, a given QUBITS-LIST (a b c ... y z) of length N would have a labeling of:

    a -> N - 1
    b -> N - 2
    c -> N - 3
    ...
    y -> 1
    z -> 0.

Qubits not present in QUBITS-LIST are mapped to themselves."
  (let ((map
          (loop :for l_i :in qubits-list
                :for p_i :from (1- (length qubits-list)) :downto 0
                :collect (cons l_i p_i))))
    (values (lambda (logical-qubit)
              (or (cdr (assoc logical-qubit map))
                  logical-qubit))
            (lambda (physical-qubit)
              (or (car (rassoc physical-qubit map))
                  physical-qubit)))))
```

We could do away with constructing `map` and just iterating, bumping indexes and the like each time.

```
(defun standard-qubit-relabeler* (qubits-list)
  "Construct a standard qubit relabeler and its inverse.

A *relabeler* is a function that takes as input a logical qubit index and produces a relabeled (sometimes called *physical*) qubit index.

The *standard labeling* of qubits (in a context where qubits are supplied as arguments to a gate application) is in descending order to 0. As such, a given QUBITS-LIST (a b c ... y z) of length N would have a labeling of:

    a -> N - 1
    b -> N - 2
    c -> N - 3
    ...
    y -> 1
    z -> 0.

Qubits not present in QUBITS-LIST are mapped to themselves."
  (let ((length (length qubits-list)))
    (values (lambda (logical-qubit)
              (let ((p (position logical-qubit qubits-list)))
                (if p
                    (- length p 1)
                    logical-qubit)))
            (lambda (physical-qubit)
              (if (<= 0 physical-qubit (1- length))
                  (nth physical-qubit qubits-list)
                  physical-qubit)))))
```

stylewarning replied:
If we do this, then the `:relabeling` argument can just be `:relabeler` and by default be `#'identity`.

ecpeterson replied:
i like this advice

notmgsk commented:
@ecpeterson @stylewarning do you know why it's not picking up the function `standard-qubit-relabeler`? I moved it into `src/relabeling.lisp`, added that to `cl-quil.asd` and `src/package.lisp` ... but it remains undefined?

notmgsk commented:
> @ecpeterson @stylewarning do you know why it's not picking up the function `standard-qubit-relabeler`? I moved it into `src/relabeling.lisp`, added that to `cl-quil.asd` and `src/package.lisp` ... but it remains undefined?

Missing a `(in-package :cl-quil)`.

stylewarning commented on src/compressor/compressor.lisp:
```diff
@@ -403,7 +401,7 @@ other's."
 (defun decompile-instructions-in-context (instructions chip-specification context)
   "This routine is called by COMPRESS-INSTRUCTIONS-IN-CONTEXT to make a decision about how to prefer 'linear algebraic compression': the list of INSTRUCTIONS can always be rewritten as its associated action matrix, but under certain conditions (governed by CONTEXT) we can sometimes get away with something less."
-  (let ((qubits-on-obj (qubits-in-instr-list instructions)))
+  (let* ((qubits-on-obj (qubits-in-instr-list instructions)))
```
```suggestion
  (let ((qubits-on-obj (qubits-in-instr-list instructions)))
```

stylewarning commented on src/compilers/approx.lisp:
```diff
@@ -714,7 +714,7 @@ Additionally, if PREDICATE evaluates to false and *ENABLE-APPROXIMATE-COMPILATIO
   (multiple-value-bind (ua ub fidelity)
       (match-matrix-to-an-e-basis-diagonalization
-       (make-matrix-from-quil center-circuit :relabeling `((,q1 . 1) (,q0 . 0)))
+       (make-matrix-from-quil center-circuit :relabeling (standard-qubit-relabeler `(,q1 ,q0)))
```
wait since it's q1 -> 1, and q0 -> 0, shouldn't it be `(list q0 q1)` ?

stylewarning replied:
also don't use backquote here it's not code

notmgsk replied:
it's reversed in standard-qubit-relabeler

stylewarning commented on src/relabeling.lisp:
```diff
@@ -0,0 +1,26 @@
+;;;; relabeling.lisp
```
idk if a new file was necessary but it's ok

ecpeterson commented on src/gates.lisp:
```diff
@@ -569,8 +569,8 @@ Note that this is a controlled version of a R_z gate multiplied by a phase."
-(defun make-matrix-from-quil (instruction-list &key relabeling)
-  "If possible, create a matrix out of the instructions INSTRUCTION-LIST, within the context of the environment ENVIRONS. If one can't be created, then return NIL.
+(defun make-matrix-from-quil (instruction-list &key (relabeling #'identity))
+  "If possible, create a matrix out of the instructions INSTRUCTION-LIST using the optional function RELABELLING that maps an input qubit index to an output qubit index. If one can't be created, then return NIL.
```
I think "relabelling" and "relabeling" are both acceptable English, but since the key arg is `relabeling`, let's stick to that.

ecpeterson commented:
Pending the resolution of Robert's comments, this looks OK to me.
